source/common/http/codec_client.cc

```diff
@@ -125,19 +125,33 @@ void CodecClient::onReset(ActiveRequest& request, StreamResetReason reason) {
 }
 
 void CodecClient::onData(Buffer::Instance& data) {
-  const Status status = codec_->dispatch(data);
-
-  if (!status.ok()) {
-    ENVOY_CONN_LOG(debug, "Error dispatching received data: {}", *connection_, status.message());
-    close();
+  bool redispatch;
+  do {
+    redispatch = false;
+    const Status status = codec_->dispatch(data);
+
+    if (!status.ok()) {
+      ENVOY_CONN_LOG(debug, "Error dispatching received data: {}", *connection_, status.message());
+      close();
+
+      // Don't count 408 responses where we have no active requests as protocol errors
+      if (!isPrematureResponseError(status) ||
+          (!active_requests_.empty() ||
+           getPrematureResponseHttpCode(status) != Code::RequestTimeout)) {
+        host_->cluster().stats().upstream_cx_protocol_error_.inc();
+      }
+    }
 
-    // Don't count 408 responses where we have no active requests as protocol errors
-    if (!isPrematureResponseError(status) ||
-        (!active_requests_.empty() ||
-         getPrematureResponseHttpCode(status) != Code::RequestTimeout)) {
-      host_->cluster().stats().upstream_cx_protocol_error_.inc();
+    // The HTTP/1 codec will pause dispatch after a single response is complete. We redispatch if
+    // there is more data.
+    // TODO(asraa): Check that there are more active requests? This causes a failure in a response
+    // with 204 integration test.
```

Comment:
Is it possible for data to be non-empty at this point when proxying H2 or H3 requests?  I think the answer is no.

Reply:
I think the answer is no as well. If all of the data was not consumed it means codec had encountered an error.

Reply:
We could add an ASSERT and see if fuzzing turns something up.

Reply:
that's correct re H2. even if the connection closes, nghttp2 will still consume any unconsumed data.

about H3, I hope it also does this. It is opaque to me, but I can guess that it does because it currently loops through all available data until there is none. i don't know about connection closing.

Reply:
I think there are some conditions about remaining data. A lot of tests fail when I add the ASSERT. Looking into it.